### ISSUE TITLE
Two-finger swipe to move through history doesn't always work

### ISSUE THREAD

**Issue by rien333** [SoongNoonien/plattenalbum#145]
# Description

I vaguely remember you never using a trackpad, so I suppose you have just never noticed this. 

In the album tracklistview, you can perform a two-finger swipe to move back to an artist's album list (pretty nice feature, IMO):

[Screencast from 2024-07-20 19-45-57.webm](https://github.com/user-attachments/assets/2a33f986-22d9-4e11-85b4-62c03c623cdc)

# To reproduce

However, aforementioned swiping motion only works under two conditions:

1. When your pointer is situated in this region:

![Screenshot from 2024-07-20 19-48-09](https://github.com/user-attachments/assets/4d76a174-5fbb-471c-98e4-b8967d67c152)

2. When the album in question only has a limited number of tracks (e.g. just two, as in the video above).

In other words, the issue is only reproducible if the users pointer is situated on the album's tracklist, and if the tracklist is sufficiently long. Not yet sure what 'sufficiently long' is here, but it could be "long enough to require scrolling". 

Happy to test any patches, if you do not have acces to a trackpad!

**Comment by SoongNoonien** [comment]
> I vaguely remember you never using a trackpad, so I suppose you have just never noticed this.

Yes, I don't have one. I only have desktop computers. :-)

This issue looks weird... The swipe gesture is a builtin feature provided by libadwaita so I have little control over it. I think what is happening when you swipe with two fingers and seemingly nothing happens is that GTK performs a horizontal scroll action in the songs list. And since this list never becomes horizontally scrollable nothing actually happens.

Can you check if the problem is somehow reproducible in the about dialog? If you click on the "Credits" button you should also be able to use a two-finger swipe to get back.

**Comment by rien333** [comment]
> Can you check if the problem is somehow reproducible in the about dialog? If you click on the "Credits" button you should also be able to use a two-finger swipe to get back.

The swipe does _not_ work in the "Credits" section of the about dialogue, but it does work in the section titled "Legal". The go- back-in-history button on my mouse works in both cases, though. 

More interestingly, the swiping motion works in every section of [Fragments](https://gitlab.gnome.org/World/Fragments)' about dialogue, including the "Credits" section, which seemingly uses the exact same UI widgetry as plattenalbum does. 

So, it would be interestingly to see ever there are any code differences between  Fragments' implementation of the About dialogue and yours. I checked other Gnome apps, and got mixed results. Most of the time swiping works correctly in about dialogues, with the exception of the Credits section in [Papers](https://gitlab.gnome.org/GNOME/Incubator/papers).

**Comment by SoongNoonien** [comment]
> The swipe does not work in the "Credits" section of the about dialogue, but it does work in the section titled "Legal". The go- back-in-history button on my mouse works in both cases, though.

This was to be expected as the "Legal" section is not scrollable.

> More interestingly, the swiping motion works in every section of Fragments' about dialogue, including the "Credits" section, which seemingly uses the exact same UI widgetry as plattenalbum does.

How did you install Fragments? Maybe this is a problem with a specific GTK/libadwaita version?

> So, it would be interestingly to see ever there are any code differences between Fragments' implementation of the About dialogue and yours.

Both use the normal `Adw.AboutDialog` there shouldn't be any difference.

> I checked other Gnome apps, and got mixed results. Most of the time swiping works correctly in about dialogues, with the exception of the Credits section in Papers.

This is another hint for a GTK/libadwaita bug in a specific version. Papers is a relatively new program which depends on very recent versions of GTK/libadwaita. How did you install Papers?

**Comment by rien333** [comment]
I install all my packages from Arch's main repo's and the AUR. Fragments is in the `extra/` repo (PKGBUILD [here](https://gitlab.archlinux.org/archlinux/packaging/packages/fragments/-/blob/main/PKGBUILD?ref_type=heads)). Papers is in the AUR, and seems to [build directly from master](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=papers).

> Maybe this is a problem with a specific GTK/libadwaita version?

Could be, actually. Just noticed that Fractal (GTK4 matrix app) has the same problem in the "Credits" section.

**Comment by SoongNoonien** [comment]
> Question is: can apps depend on different libadwaita versions, even though I'm not using anything like Flatpak?

Yes, they could. But not in arch ;-) So, I have no idea why these apps behave so differently. Just to be sure, you did fiddle around with the cursor placement and so on in every app? And you are certain that some apps have this problem and others definitely don't?

**Comment by rien333** [comment]
> And you are certain that some apps have this problem and others definitely don't?

Okay, I think I figured out this part of the issue. The reason  swiping doesn't work in the "Credits" section of the about dialogue in some apps, has to do with the number of items in that list. If the list is large enough to spawn a scrollbar, the swipe stops working. If I resize the dialogue window to a point where the scrollbar disappears,  the swipe works again.

Hence, it seems like your initial hypothesis about the tracklist "stealing" the swipe events is on the right track. 

Is this worth reporting to the adwaita devs, if nobody has already done so?

**Comment by SoongNoonien** [comment]
> Is this worth reporting to the adwaita devs, if nobody has already done so? Seems to me that being able to use history swiping gestures and lists at the same time is desirable.

Definitely! There is a libadwaita demo app which incidentally has an about dialog of its own. If you are able to reproduce it there you can easily make a bug report without the need to write a reproducer of your own. In arch the package is named `libadwaita-demos` and the respective command to run the demo is `adwaita-1-demo`.

**Comment by rien333** [comment]
Thanks for the pointers!

Here is the upstream issue, for posterity's sake: https://gitlab.gnome.org/GNOME/libadwaita/-/issues/909
